modules/standard/Heap.chpl

```diff
@@ -0,0 +1,400 @@
+    */
+  }
+    
+  record heap {
+
+    /* The type of the elements contained in this heap. */
+    type eltType;
+
+    /* If `true`, this heap will perform parallel safe operations. */
+    param parSafe = false;
+
+    /*
+      Comparator record that defines how the
+      data is compared. The greatest element will be on the top.
+    */
+    var comparator: record;
+
+    pragma "no doc"
+    var _lock$ = if parSafe then new _LockWrapper() else none;
+
+    /*
+      Use a list to store elements.
+    */
+    pragma "no doc"
+    var _data: list(eltType);
+
+    /*
+      Build the heap from elements that have been stored, from bottom to top
+      in O(N)
+    */
+    pragma "no doc"
+    proc _commonInitFromIterable(iterable)
+    lifetime this < iterable {
+      _data = new list(eltType);
+      for x in iterable do
+        _data.append(x);
+      for i in 0 .. _data.size-1 by -1 {
+        _heapify_down(i);
+      }
+    }
+
+    /*
+      Initializes an empty heap.
+
+      :arg eltType: The type of the elements
+
+      :arg comparator: `DefaultComparator` makes max-heap and `ReverseCompartor` makes a min-heap
+
+      :arg parSafe: If `true`, this heap will use parallel safe operations.
+      :type parSafe: `param bool`
+    */
+    proc init(type eltType, param parSafe = false, comparator: record = defaultComparator) {
+      _checkType(eltType);
+      this.eltType = eltType;
+      this.parSafe = parSafe;
+      this.comparator = comparator;
+      this._data = new list(eltType);
+    }
+
+    /*
+      Initializes a heap containing elements that are copy initialized from
+      the elements contained in another heap.
+
+      :arg other: The heap to initialize from.
+    */
+    proc init=(other: heap(this.type.eltType)) {
+      _checkType(this.type.eltType);
+      if !isCopyableType(this.type.eltType) then
+        compilerError("Cannot copy heap with element type that cannot be copied");
+
+      this.eltType = this.type.eltType;
+      this.parSafe = this.type.parSafe;
+      this.comparator = other.comparator;
+      this.complete();
+      _commonInitFromIterable(other._data);
+    }
+
+    /*
+      Locks operations
+    */
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+
+    /*
+      Return the size of the heap.
+
+      :return: The size of the heap
+      :rtype: int
+    */
+    proc size: int {
+      _enter();
+      var result = _data.size;
+      _leave();
+      return result;
+    }
+
+    /*
+      Returns `true` if this heap contains zero elements.
+
+      :return: `true` if this heap is empty.
+      :rtype: `bool`
+    */
+    proc isEmpty(): bool {
+      _enter();
+      var result = _data.isEmpty();
+      _leave();
+      return result;
+    }
+
+    /*
+      Return the maximal element in the heap.
```

Comment:
Actually, I think *maximal* can describe the default behavior. Passing a default comparator and users can get a maximal element. If users want minimal, then just reverse the comparator. When saying `top`, I can't tell the heap will return a maximal one or a minimal one.

Reply:
>  When saying top, I can't tell the heap will return a maximal one or a minimal one.

Right, and I want the document to be ambiguous there. For example, if I create a heap with reverse comparator, calling `top` will give me the minimum element in the heap. But the documentation says "maximal" and it may be confusing

Reply:
https://github.com/chapel-lang/chapel/pull/15997/commits/acfb880ea0bcd8b2418361d1f7b38c2a450fb08d What do you think of this?

Reply:
Looks good